---
title: "Data Processing (fastq to bams) for sequencing data produced by Sard and Homola"
author: "Nicole Adams"
date: "2025-03-02"
output: 
  html_document:
    toc: true
    code_folding: show
---

```{r setup, include=FALSE, echo=TRUE}
# Load necessary packages
library(tidyverse)
library(data.table)
library(viridisLite)
library(readxl)
library(googlesheets4) #access Google drive with RSC sample data see https://www.tidyverse.org/blog/2020/05/googlesheets4-0-2-0/
library(lubridate)
library(kableExtra)
library(ggrepel)
library(ggpubr)

# Define Rmd knit options
knitr::opts_knit$set(root.dir = "..")
knitr::opts_chunk$set(
	fig.height = 6,
	fig.width = 8,
	warning = FALSE, # suppress warnings
	message = FALSE # suppress messages
)
#echo = FALSE # defaults to not showing R code, change chunk to echo=TRUE where wanted

# Set eval=FALSE for all bash chunks
knitr::opts_hooks$set(
  engine = function(options) {
    if (options$engine == "bash") {
      options$eval <- FALSE
    }
    options
  }
)
```

# Processing raw RAD capture sequence data for Red Swamp Crayfish<br> 
Using Nick Sard's raw data (collected in 2018) and Jared Homola's raw data (collected in 2019-2020).
This work flow is for bioinformatic re-processing of invasive Red Swamp Crayfish (P. clarkii) samples originally sequenced in 2018 by Nick Sard and a second sequencing run in 2021 by Jared Homola.

&nbsp;

## Download the new RSC genome to MSU HPCC
[RSC genome](https://www.ncbi.nlm.nih.gov/genome/?term=txid6728[orgn]) 
```{bash, eval=FALSE}
cd /mnt/home/adamsn23/RedSwampCrayfish/reference 

wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/020/424/385/GCF_020424385.1_ASM2042438v2/GCF_020424385.1_ASM2042438v2_genomic.fna.gz 
 
```

&nbsp;

## Index the reference genome and make a directory
(/mnt/home/adamsn23/indexRef_RSC.sbatch) 
```{bash, eval=FALSE}
#!/bin/sh 
#SBATCH --ntasks=1 
#SBATCH -t 12:00:00 
#SBATCH -N 1 -c 8 
#SBATCH --mem 16G 
#SBATCH --job-name index 
#SBATCH --output=index.%j.out 
#SBATCH --error=index.%j.err 
#SBATCH --mail-type=END 
#SBATCH --mail-user=username@msu.edu 
 
module load GCC/6.4.0-2.28 OpenMPI/2.1.1 BWA/0.7.17 SAMtools/1.9 
module load picard/2.25.0-Java-11 
 
cd /mnt/home/adamsn23/RedSwampCrayfish/reference/ 
 
gunzip GCF_020424385.1_ASM2042438v2_genomic.fna.gz  
 
bwa index -a bwtsw GCF_020424385.1_ASM2042438v2_genomic.fna 
 
samtools faidx GCF_020424385.1_ASM2042438v2_genomic.fna 
 
java -jar $EBROOTPICARD/picard.jar CreateSequenceDictionary R=GCF_020424385.1_ASM2042438v2_genomic.fna O=GCF_020424385.1_ASM2042438v2_genomic.fna.dict 
 
gzip GCF_020424385.1_ASM2042438v2_genomic.fna 
```

&nbsp;

## Pre-processing needed for flipReads
```{bash, eval=FALSE}
# Create directories needed for 2_flipReads_NEA.sh
mkdir /mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/QSTAT
mkdir /mnt/gs18/scratch/users/adamsn23/RSC

# Fix barcode file names to match fastq file names
#rename Jared's barcodes with leading 0s (01-09)
mv RSC01_barcodes.txt RSC1_barcodes.txt 

#rename Nick's barcodes replace PCL1L1 with RSCL1 and add 'nicks' so I know where it came from 
mv PCL1L1.barcodes.and.ids.txt RSCL1_nicks.barcodes.and.ids.txt 
```

&nbsp;

## Unzip and flip reads for both Jared's and Nick's fastqs.  
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/2_flipReads_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: Flip_reads.sh
#   Date: 01/05/21, 02/03/2022 (NEA)
#   Description: Flip the orientation of the Best RAD reads
#--------------------------------------------------------------------------------------------------
#       Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
jDATA_PATH=/mnt/research/Scribner_Lab/projects/RedSwampCrayfish_MISGP
nDATA_PATH=/mnt/research/Scribner_Lab/projects/Red_Swamp_Crayfish_RAD_2018
my_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

mkdir $my_PATH/SHELL/flipReads
mkdir $my_PATH/QSTAT/flipReads

mkdir $SCR/RSC/unzipped/
mkdir $SCR/RSC/flipped/ #this line wasn't in Jared's script


### Unzip the files - Jared's
cd $my_PATH/SHELL/flipReads

find $jDATA_PATH/RAW -name '*.fq.gz' | while read -r LINE
do
dat=$(echo $LINE | awk -F "/" '{print $10}' | awk -F ".fq" '{print $1}')
datn=$(echo $LINE | awk -F "/" '{print $10}' | awk -F "_" '{print $1}') 

echo '#!/bin/sh 
#SBATCH -N 1
#SBATCH -t 1:00:00
#SBATCH --mem 64G 
#SBATCH -J '$dat'_unzip
#SBATCH -o '$my_PATH'/QSTAT/flipReads/'$dat'_unzip.o
#SBATCH --error='$my_PATH'/QSTAT/flipReads/'$dat'_unzip.err

## Unzip
gunzip -c '$LINE' > '$SCR'/RSC/unzipped/'$dat'.fq

scontrol show job ${SLURM_JOB_ID} $' > "$dat"_unzip.sh

sbatch "$dat"_unzip.sh

done


### Unzip the files - Nick's
find $nDATA_PATH/all_raw -name '*.fastq.gz' | while read -r LINE2
do
dat2=$(echo $LINE2 | awk -F "/" '{print $8}' | awk -F ".fastq" '{print $1}')
dat2n=$(echo $LINE2 | awk -F "/" '{print $8}' | awk -F "_" '{print $1}') 

echo '#!/bin/sh 
#SBATCH -N 1
#SBATCH -t 1:00:00
#SBATCH --mem 64G 
#SBATCH -J '$dat2'_unzip
#SBATCH -o '$my_PATH'/QSTAT/flipReads/'$dat2'_unzip.o
#SBATCH --error='$my_PATH'/QSTAT/flipReads/'$dat2'_unzip.err

## Unzip
gunzip -c '$LINE2' > '$SCR'/RSC/unzipped/'$dat2'.fastq

scontrol show job ${SLURM_JOB_ID} $' > "$dat2"_unzip.sh

sbatch "$dat2"_unzip.sh

done


#### Flip reads - Jared's

cd $my_PATH/SHELL/flipReads

find $jDATA_PATH/RAW -name '*1.fq.gz' | while read -r LINE
do
dat=$(echo $LINE | awk -F "/" '{print $10}' | awk -F ".fq" '{print $1}')
datn=$(echo $LINE | awk -F "/" '{print $10}' | awk -F "_" '{print $1}') 
num=$(echo $LINE | awk -F "/" '{print $10}' | awk -F "-" '{print $3}' | awk -F "_" '{print $1}')

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 12:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=64G
#SBATCH -J '$datn'_flipReads
#SBATCH -o '$my_PATH'/QSTAT/flipReads/'$datn'_flipReads.o
#SBATCH --error='$my_PATH'/QSTAT/flipReads/'$datn'_flipReads.err

perl '$my_PATH'/SHELL/dependencies/bRAD_flip_trim.pl \
    '$my_PATH'/SHELL/dependencies/'$datn'_barcodes.txt \
    '$SCR'/RSC/unzipped/'$datn'_CKDL200169916-1a-'$num'_HNNJHDSXY_L2_1.fq \
    '$SCR'/RSC/unzipped/'$datn'_CKDL200169916-1a-'$num'_HNNJHDSXY_L2_2.fq \
    '$SCR'/RSC/flipped/'$datn'_flipped.1.fq \
    '$SCR'/RSC/flipped/'$datn'_flipped.2.fq
    

scontrol show job ${SLURM_JOB_ID} $' > "$datn"_flipReads.sh
sbatch "$datn"_flipReads.sh

done


#### Flip reads - Nick's

cd $my_PATH/SHELL/flipReads

find $nDATA_PATH/all_raw -name '*R1_001.fastq.gz' | while read -r LINE2
do
dat2=$(echo $LINE2 | awk -F "/" '{print $8}' | awk -F ".fastq" '{print $1}')
dat2n=$(echo $LINE2 | awk -F "/" '{print $8}' | awk -F "_" '{print $1}') 
nam=$(echo $LINE2 | awk -F "/" '{print $8}' | awk -F "_R1_001" '{print $1}') 

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 12:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=64G
#SBATCH -J '$dat2n'_flipReads
#SBATCH -o '$my_PATH'/QSTAT/flipReads/'$dat2n'_flipReads.o
#SBATCH --error='$my_PATH'/QSTAT/flipReads/'$dat2n'_flipReads.err

perl '$my_PATH'/SHELL/dependencies/bRAD_flip_trim.pl \
    '$my_PATH'/SHELL/dependencies/'$dat2n'_nicks.barcodes.and.ids.txt \
    '$SCR'/RSC/unzipped/'$nam'\_R1_001.fastq \
    '$SCR'/RSC/unzipped/'$nam'\_R2_001.fastq \
    '$SCR'/RSC/flipped/'$dat2n'_flipped.1.fq \
    '$SCR'/RSC/flipped/'$dat2n'_flipped.2.fq

scontrol show job ${SLURM_JOB_ID} $' > "$dat2n"_flipReads.sh
sbatch "$dat2n"_flipReads.sh

done

```

&nbsp;

### Zip the flipped reads
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/3_zipFlipped_NEA.sh )
```{bash, eval=FALSE}
#==================================================================================================
#   File: Zip_flipped.sh
#   Date: 01/06/21, 02/07/2022 (NEA)
#   Description: Zip the flipped best RAD reads
#--------------------------------------------------------------------------------------------------
#       Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
my_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

cd $my_PATH/SHELL/flipReads

ls $SCR/RSC/flipped | grep -v ".2.fq" | awk -F "_flipped" '{print $1}' | while read -r LINE
do

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 3:59:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=64G
#SBATCH -J Zip_'$LINE'
#SBATCH -o '$my_PATH'/QSTAT/flipReads/Zip_'$LINE'.o
#SBATCH --error '$my_PATH'/QSTAT/flipReads/Zip_'$LINE'.err

cd '$SCR'/RSC/flipped

gzip < ./'$LINE'_flipped.1.fq > '$LINE'_flipped.1.fq.gz
gzip < ./'$LINE'_flipped.2.fq > '$LINE'_flipped.2.fq.gz

scontrol show job ${SLURM_JOB_ID} $' > Zip_"$LINE".sh

sbatch Zip_"$LINE".sh

done

```

&nbsp;

## Demultiplex
### Pre-processing for demultiplexing
Change name of Nick's barcode file again to match demultiplex script
```{bash, eval=FALSE}
cp SHELL/dependencies/RSCL1_nicks.barcodes.and.ids.txt SHELL/dependencies/RSCL1_barcodes.txt  
cp SHELL/dependencies/RSCL2_nicks.barcodes.and.ids.txt SHELL/dependencies/RSCL2_barcodes.txt 
cp SHELL/dependencies/RSCL3_nicks.barcodes.and.ids.txt SHELL/dependencies/RSCL3_barcodes.txt 
 
```

&nbsp;

### Demultiplex
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/4_demultiplex_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: 4_demultiplex.sh
#   Date: 01/06/21,  02/08/2022 (NEA)
#   Description: Demultiplex libraries using process_radtags
#--------------------------------------------------------------------------------------------------
#   Author: Jared Homola, Nicole Adams
#==================================================================================================

RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

mkdir $RUN_PATH/SHELL/demult
mkdir $RUN_PATH/QSTAT/demult
mkdir $SCR/RSC/demult

cd $RUN_PATH/SHELL/demult

### Demultiplex 
ls $SCR/RSC/flipped | grep -v ".gz" | grep -v ".2.fq" | awk -F "_flipped" '{print $1}' | while read -r LINE
do

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 3:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=24G
#SBATCH -J '$LINE'_demultiplex
#SBATCH -o '$RUN_PATH'/QSTAT/demult/'$LINE'_demultiplex.o
#SBATCH --error '$RUN_PATH'/QSTAT/demult/'$LINE'_demultiplex.err

module purge
module load GCC/9.3.0 OpenMPI/4.0.3 Stacks/2.59

process_radtags \
    -1 '$SCR'/RSC/flipped/'$LINE'_flipped.1.fq.gz \
    -2 '$SCR'/RSC/flipped/'$LINE'_flipped.2.fq.gz \
    -i gzfastq \
    -y gzfastq \
    -o '$SCR'/RSC/demult/ \
    -b '$RUN_PATH'/SHELL/dependencies/'$LINE'_barcodes.txt \
    --inline_null \
    -e sbfI \
    --barcode_dist_1 1 \
    --retain_header

scontrol show job ${SLURM_JOB_ID}' > demultiplex_"$LINE".sh

sbatch demultiplex_"$LINE".sh

done
```

&nbsp;

## Remove clones
Some debate on whether or not to remove clones. I found [this stacks tutorial](https://gist.github.com/Astahlke/e07e346720061164855fb2c2a7dc7f02) that said the following:<br>  
>PCR could stochastically amplify one of the alleles more than the other, sometimes leading to individuals appearing as homozygotes rather than heterozygotes (Andrews et al 2016). For this reason, removing PCR duplicates likely reduces false locus assembly and the variance in number of loci across individuals. Downstream, this can have differing effects on population structure depending on the dataset (Díaz-Arce & Rodríguez-Ezpeleta 2019). However, this may be a negligible effect (Euclide et al. 2019).<br>

(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/5_removeClones_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: removeClones.sh
#   Date: 01/06/21, 02/08/2022 (NEA)
#   Description: Filter clonal reads from data
#--------------------------------------------------------------------------------------------------
#	Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

mkdir $SCR/RSC/cloneFilter
mkdir $RUN_PATH/SHELL/cloneFilter
mkdir $RUN_PATH/QSTAT/cloneFilter

cd $RUN_PATH/SHELL/cloneFilter

# too many files to submit to slurm so did PC and FC separate, then did the rest
ls $SCR/RSC/demult/ | grep "fq" | grep -v ".2.fq.gz" | grep -v "rem" | grep -v "PC" | grep -v "FC" | sed 's/\.1\.fq\.gz//g' | sort | uniq | while read -r LINE
do

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 1:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=32G
#SBATCH -J '$LINE'_cloneFilter
#SBATCH -o '$RUN_PATH'/QSTAT/cloneFilter/cloneFilter.'$LINE'.o
#SBATCH --error '$RUN_PATH'/QSTAT/cloneFilter/cloneFilter.'$LINE'.err

module purge
module load GCC/9.3.0 OpenMPI/4.0.3 Stacks/2.59

clone_filter -1 '$SCR'/RSC/demult/'$LINE'.1.fq.gz \
    -2 '$SCR'/RSC/demult/'$LINE'.2.fq.gz \
    -i gzfastq \
    -o '$SCR'/RSC/cloneFilter/

scontrol show job ${SLURM_JOB_ID}' > ./cloneFilter."$LINE".sh

sbatch ./cloneFilter."$LINE".sh

done


### Summarize results
##grep '% clone reads' $RUN_PATH/QSTAT/cloneFilter/cloneFilter*.err >> $RUN_PATH/OUT/cloneFilterRes.txt

#for FILE in QSTAT/cloneFilter/cloneFilter*.err 
#do 
#ind=$(echo $FILE | awk -F "/" '{print $3}' | awk -F "." '{print $2}')  
##perc=$(grep -oP '.{0,5}%' $FILE) 
#res=$(grep '% clone'  $FILE | grep -Eo '[0-9]{1,}' | tr -s '\n' '\t') 
#echo $ind $res >> OUT/cloneFilterRes.txt 
#done 
```

&nbsp;

## Trim and filter
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/6_trimQualityFilter_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: 6_trimQualityFilter.sh
#   Date: 01/07/20, 02/10/2022 (NEA)
#   Description: Use Trimmomatic to trim reads and quality filter
#--------------------------------------------------------------------------------------------------
#	Authors: Jared Homola, Nicole Adams
#==================================================================================================

RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

### Need to rename files by removing the extra .1 or .2 that was inserted
find $SCR/RSC/cloneFilter -type f -name '*.1.1.fq.gz' | while read FILE ; do
    newfile="$(echo ${FILE} | sed -e 's/\.1.fq.gz/.fq.gz/')" ;
    mv "${FILE}" "${newfile}" ;
done 

find $SCR/RSC/cloneFilter -type f -name '*.2.2.fq.gz' | while read FILE ; do
    newfile="$(echo ${FILE} | sed -e 's/\.2.fq.gz/.fq.gz/')" ;
    mv "${FILE}" "${newfile}" ;
done 


mkdir $SCR/RSC/trimmed
mkdir $RUN_PATH/SHELL/trimQualityFilter
mkdir $RUN_PATH/QSTAT/trimQualityFilter

cd $RUN_PATH/SHELL/trimQualityFilter

ls $SCR/RSC/cloneFilter | grep "fq" | grep -v ".2.fq.gz" | grep "FC" | sed 's/\.1\.fq\.gz//g' | while read -r LINE
do
echo '#!/bin/sh 
#SBATCH --nodes=1-4
#SBATCH --ntasks=1
#SBATCH -t 3:00:00
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=8G
#SBATCH -J '$LINE'.trimQualityFilter
#SBATCH -o '$RUN_PATH'/QSTAT/trimQualityFilter/trimQualityFilter.'$LINE'.o
#SBATCH --error '$RUN_PATH'/QSTAT/trimQualityFilter/trimQualityFilter.'$LINE'.err

module purge
module load Trimmomatic/0.39-Java-11 Java/1.8.0_162

cd '$SCR'/RSC/cloneFilter
java -jar $EBROOTTRIMMOMATIC/trimmomatic-0.39.jar PE -threads 4 -phred33 \
'$LINE'.1.fq.gz '$LINE'.2.fq.gz \
'$SCR'/RSC/trimmed/'$LINE'_paired.1.fq.gz '$SCR'/RSC/trimmed/'$LINE'_unpaired.1.fq.gz \ '$SCR'/RSC/trimmed/'$LINE'_paired.2.fq.gz '$SCR'/RSC/trimmed/'$LINE'_unpaired.2.fq.gz \
ILLUMINACLIP:/mnt/research/Scribner_Lab/projects/RedSwampCrayfish_MISGP/SHELL/dependencies/adapters.fa:2:30:10 SLIDINGWINDOW:4:15 MINLEN:50 

scontrol show job ${SLURM_JOB_ID}' > ./trimQualityFilter."$LINE".sh

sbatch ./trimQualityFilter."$LINE".sh

done


###### How many reads were trimmed and filtered? ######
#grep 'Surviving' $RUN_PATH/QSTAT/trimQualityFilter/trimQualityFilter*.o >> $RUN_PATH/OUT/trimQualityFilterRes.txt

#for FILE in QSTAT/trimQualityFilter/trimQualityFilter*.err
#do
#ind=$(echo $FILE | awk -F "/" '{print $3}' | awk -F "." '{print $2}')
#res=$(grep 'Surviving'  $FILE | awk -F " " '{print $4 "\t" $7 "\t" $8 "\t" $12 "\t" $13 "\t" $17 "\t" $18 "\t" $20 "\t" $21}') 
#echo $ind $res >> OUT/trimQualityFilterRes.txt
#done


```

&nbsp;

## Mapping to reference
### Mapping pre-processing
```{bash, eval=FALSE}
mkdir $SCRATCH/RSC/trimmed/paired

mv $SCRATCH/RSC/trimmed/*_paired* $SCRATCH/RSC/trimmed/paired

```

&nbsp;

### map
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/8_bwaMemMap_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: bwaMemMap.sh
#   Date: 01/07/21, 02/11/2022 (NEA)
#   Description: Map reads to reference genome using bwa-mem
#--------------------------------------------------------------------------------------------------
#	Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

mkdir $SCR/RSC/mapped
mkdir $RUN_PATH/SHELL/mapping
mkdir $RUN_PATH/QSTAT/mapping

cd $RUN_PATH/SHELL/mapping
ls $SCR/RSC/trimmed/paired | grep "fq" | grep -v ".2.fq.gz" | sed 's/\_paired.1\.fq\.gz//g' | grep "SHD" | uniq | while read -r LINE
do

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 3:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=36G
#SBATCH -J '$LINE'.bwaMemMap
#SBATCH -o '$RUN_PATH'/QSTAT/mapping/'$LINE'.bwaMemMap.o
#SBATCH --error '$RUN_PATH'/QSTAT/mapping/'$LINE'.bwaMemMap.err

module purge
module load GCC/6.4.0-2.28 OpenMPI/2.1.1 BWA/0.7.17 SAMtools/1.9

cd '$SCR'/RSC/trimmed/paired

bwa mem -R "@RG\tID:'$LINE'\tSM:'$LINE'\tPL:ILLUMINA\tLB:LB1" /mnt/home/adamsn23/RedSwampCrayfish/reference/GCF_020424385.1_ASM2042438v2_genomic.fna ./'$LINE'_paired.1.fq.gz ./'$LINE'_paired.2.fq.gz | samtools view -Sb - > ../../mapped/'$LINE'.pe.bam 

scontrol show job ${SLURM_JOB_ID}' > ./bwaMemMap."$LINE".sh

sbatch ./bwaMemMap."$LINE".sh

done


### Log results of mapping

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis

cd $RUN_PATH/SHELL

echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 1:30:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=12G 
#SBATCH -J bwaFlagstats
#SBATCH -o '$RUN_PATH'/QSTAT/bwaFlagstats.o
#SBATCH -err '$RUN_PATH'/QSTAT/bwaFlagstats.err

module purge
module load GCC/8.3.0 SAMtools/1.10

cd '$SCR'/RSC/mapped

ls | grep ".pe.bam" | sort | uniq | while read -r LINE
do
   echo "$LINE" >> '$RUN_PATH'/OUT/mappingResults.txt
   samtools flagstat "$LINE" >> '$RUN_PATH'/OUT/mappingResults.txt
done' > bwaFlagstats.sh

sbatch bwaFlagstats.sh




###### WITHOUT REMOVE CLONES ########
# see 8_bwaMemMap.sh

```

&nbsp;

### sort
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/10_sortBAM_NEA.sh)
```{bash, eval=FALSE}
#==================================================================================================
#   File: 10_sortBAM.sh
#   Date: 01/07/20, 02/15/2022 (NEA)
#   Description: Sort BAM files
#--------------------------------------------------------------------------------------------------
#	Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

mkdir $RUN_PATH/QSTAT/sortBAM
mkdir $RUN_PATH/SHELL/sortBAM_jobs
mkdir $RUN_PATH/OUT/sorted

cd $RUN_PATH/SHELL/sortBAM_jobs
ls $SCR/RSC/mapped | grep bam | awk -F "." '{print $1}' | grep "SHD" | sort | uniq | while read -r LINE
do
echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 1:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=8G
#SBATCH -J '$LINE'.sortBAM
#SBATCH -o '$RUN_PATH'/QSTAT/sortBAM/'$LINE'.sortBAM.o
#SBATCH --error '$RUN_PATH'/QSTAT/sortBAM/'$LINE'.sortBAM.err

module purge
module load GCC/6.4.0-2.28 OpenMPI/2.1.1 SAMtools/1.10

samtools sort '$SCR'/RSC/mapped/'$LINE'.pe.bam -o '$RUN_PATH'/OUT/sorted/'$LINE'.pe.sorted.bam

scontrol show job ${SLURM_JOB_ID}' > ./sortBAM."$LINE".sh

sbatch ./sortBAM."$LINE".sh

done

```

&nbsp;

## Filter bams
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/11_filterBAMS_NEA.sh)
-q 20=keep quality >= 20; -h=keep header; -f2=keep proper pairs(PROPER_PAIR) ; -F2308=exclude unmapped reads, not primary alignment, supplementary alignment.  see [this calculater](https://broadinstitute.github.io/picard/explain-flags.html)
```{bash, eval=FALSE}
#==================================================================================================
#   File: filterBAM.sh
#   Date: 01/07/21, 03/03/2022 (NEA)
#   Description: Filter BAM files
#--------------------------------------------------------------------------------------------------
#	Authors: Jared Homola, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

mkdir $RUN_PATH/QSTAT/filterBAM
mkdir $RUN_PATH/SHELL/filterBAM_jobs
mkdir $RUN_PATH/OUT/filtered

cd $RUN_PATH/SHELL/filterBAM_jobs
ls $RUN_PATH/OUT/sorted | grep bam | awk -F "." '{print $1}' | grep -v "FC" | grep -v "SHD" | sort | uniq | while read -r LINE
do
echo '#!/bin/sh 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH -t 1:00:00
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=8G
#SBATCH -J '$LINE'.filterBAM
#SBATCH -o '$RUN_PATH'/QSTAT/filterBAM/'$LINE'.filterBAM.o
#SBATCH --error '$RUN_PATH'/QSTAT/filterBAM/'$LINE'.filterBAM.err

module purge
module load GCC/6.4.0-2.28 OpenMPI/2.1.1 SAMtools/1.9

cd '$RUN_PATH'/OUT/sorted
samtools view -q 20 -h -f2 -F2308  '$LINE'.pe.sorted.bam | samtools view -Sb > ../filtered/'$LINE'.pe.sorted.filtered.bam

scontrol show job ${SLURM_JOB_ID}' > ./filterBAM."$LINE".sh

sbatch ./filterBAM."$LINE".sh

done
```

&nbsp;

## Mapping metrics
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/11B_mapMetrics_NEA.sh)<br>
filtered (11B_mapMetrics_filtered_NEA.sh)
```{bash, eval=FALSE}
#!/bin/sh 
#SBATCH --nodes=1-4
#SBATCH --ntasks=6
#SBATCH -t 3:00:00
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=8G
#SBATCH -J mapMetrics
#SBATCH -o /mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/QSTAT/mapMetrics.o
#SBATCH --error /mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/QSTAT/mapMetrics.err

#==================================================================================================
#   File: 11B_mapMetrics_NEA.sh
#   Date: 03/22/2022
#   Description: Get summary of mapping
#--------------------------------------------------------------------------------------------------
#       Authors: Nicole Adams
#==================================================================================================

module purge
module load GCC/6.4.0-2.28 OpenMPI/2.1.1 SAMtools/1.9

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23
BAM_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/OUT/sorted


cd $RUN_PATH/SHELL

touch $BAM_PATH/mapSummary_nickYjared.txt

for sample in `ls $BAM_PATH/*sorted.bam | cut -f1 -d'.'`
do
  samtools flagstat "$sample".pe.sorted.bam -@ 6  > "$sample"_mapSum.txt
 awk 'FNR == 1{ print FILENAME }' "$sample"_mapSum.txt >> $BAM_PATH/mapSummary_nickYjared.txt
 cat "$sample"_mapSum.txt >> $BAM_PATH/mapSummary_nickYjared.txt

done

for sample in $BAM_PATH/*mapSum.txt; do awk 'FNR == 1{ print FILENAME } {printf "%-20s %-40s\n", $1, $3}' OFS="\t" $sample | awk '
{
    for (i=1; i<=NF; i++)  {
        a[NR,i] = $i
    }
}
NF>p { p = NF }
END {
    for(j=1; j<=p; j++) {
        str=a[1,j]
        for(i=2; i<=NR; i++){
            str=str" "a[i,j];
        }
        print str
    }
}' >> $BAM_PATH/mapSummary_nickYjared2.txt; done

grep 'mnt' $BAM_PATH/mapSummary_nickYjared2.txt > $BAM_PATH/mapSummary_nickYjared3.txt

rm $BAM_PATH/*_mapSum.txt
```

&nbsp;

### Examing mapping results with R
I ran summary stats for sorted bams (.pe.sorted.bam) and filtered bams (.pe.sorted.bam)
```{r, message=FALSE, warning=FALSE}
msumA <- read.table("~/Documents/crayfish_lab/nickYjaredReanalysis/mapSummary_nickYjared3.txt")

colnames(msumA) <- c("Sample", "QCpassedReads", "secondary", "supplementary", "duplicates", "mapped", "paired", "read1", "read2", "properlyPaired", "itselfYmateMapped", "singletons", "mateMappedDiffChr", "mateMappedDiffChr_mapQ5")

nothanks <- c("a", "b", "c", "d", "e", "f", "g", "h")
msum <- msumA %>% separate(Sample, c("a", "b", "c", "d", "e", "f", "g", "h", "i"), sep = "/") %>% select(-nothanks) %>%
  separate(i, c("Sample", "stuff"), sep = "_") %>% select(-stuff)

msum$percentMap <- (msum$mapped/msum$QCpassedReads)*100
msum$percentPaired <- (msum$properlyPaired/msum$paired)*100
msum$percentSingle <- (msum$singletons/msum$properlyPaired)*100

msum.df <- msum
goby <- c("RGALP002")  #remove round goby sample :/

msum.df <- msum.df %>% filter(Sample != "RGALP002")

hist(msum.df$percentMap)

msum.p <- ggplot(msum.df, aes( percentMap)) +
    geom_histogram(color="gray40", fill="gray") +
    theme_minimal() 

msum.tab <- t(as.data.frame(c(length(msum.df$Sample), mean(msum.df$percentMap,na.rm = T), mean(msum.df$percentPaired, na.rm = T)))) 
colnames(msum.tab) <- c("Samples", "percentMap", "percentPaired")
msum.tab2 <- tibble(
    Samples = length(msum.df$Sample),
    percentMap = mean(msum.df$percentMap, na.rm = TRUE),
    percentPaired = mean(msum.df$percentPaired, na.rm = TRUE)
)

msum.tab2 %>%
  kbl(digits = 2,
    caption = "Mapping Summary",
    col.names = c("Total Samples (N)", "Mean % Mapped", "Mean % Paired"),
    format = "html"
  ) %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed"),
    full_width = FALSE,
    font_size = 14
)
```

&nbsp;

#### Identify low mapping samples
```{r, message=FALSE, warning=FALSE}
msum.low <- msum.df %>% filter(percentMap < 80) %>% select(Sample, QCpassedReads, percentPaired, percentMap)

# PCFCS004, SHD-J42, SHD-J43, SHD-J44

kable(msum.low, digits = 2, caption = "low mapping") %>% kableExtra::kable_styling()
```

&nbsp;

## Read counts
(.../RedSwampCrayfish/nickYjaredReanalysis/scripts/12_countReads_NEA.sh)
```{bash, eval=FALSE}
#!/bin/sh 
#SBATCH --nodes=1-4
#SBATCH --ntasks=8
#SBATCH -t 3:00
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=8G
#SBATCH -J Count_reads
#SBATCH -o /mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/QSTAT/Count_reads.o
#SBATCH --error /mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis/QSTAT/Count_reads.err

#==================================================================================================
#   File: Count_reads.sh
#   Date: 04/22/19, 03/11/2022 (NEA)
#   Description: Flip the orientation of the Best RAD reads
#       Run: Interactively - array
#--------------------------------------------------------------------------------------------------
#       Authors: Jared Homola, Seth Smith, Nicole Adams
#==================================================================================================

#Define alias for project root directory
RUN_PATH=/mnt/home/adamsn23/RedSwampCrayfish/nickYjaredReanalysis
SCR=/mnt/gs18/scratch/users/adamsn23

cd $RUN_PATH/SHELL

ls $SCRATCH/RSC/trimmed/paired | grep "fq" | grep -v ".2.fq.gz" | grep -v "rem" | awk -F "." '{print $1}' | sort | uniq | while read -r LINE 
do 
ind=$(echo $LINE) 
dat=$(zcat $SCRATCH/RSC/trimmed/paired/$LINE.1.fq.gz | echo $((`wc -l`/4))) 
echo $ind $dat >> OUT/Read_counts.txt 
done  

```

&nbsp;

### Examine read counts in R
```{r, message=FALSE, warning=FALSE}
counts.a <- as.data.frame(read.delim("~/Documents/crayfish_lab/nickYjaredReanalysis/readCounts.txt", header = FALSE))

counts.b <- counts.a %>% separate(V1, into = c("ind", "reads"), sep = " ") %>%
  separate(ind, into = c("ind", "stuff"), sep = "_") %>% select(-stuff)

counts <- counts.b %>% mutate(reads= as.numeric(reads))


# identify samples to exclude from further analyses:
qts <- quantile(counts$reads,probs=c(.05,.95))

counts.p <- ggplot(counts, aes(x=reads)) +
 # geom_density() +
  geom_histogram() +
  geom_vline(xintercept = qts[1], color="red", linetype="dashed") +
  geom_vline(xintercept = qts[2], color="red", linetype="dashed") +
  theme_minimal()

low.counts <- counts %>% filter(reads < qts[1])
low.counts %>% kbl(
  caption = "Samples with Low Read Counts", # Add a title
  col.names = c("Individual ID", "Read Count"), # Clean column names
  align = 'c',
  format = "html")  %>%
  kable_styling(
    bootstrap_options = c("striped", "hover", "condensed"), # Popular styling options
    full_width = T)
```